### PR TITLE
Telemetry improvements

### PR DESCRIPTION
**Pull request by bsudekum**
* [x] Prerequisite: https://github.com/mapbox/mapbox-navigation-ios/pull/1013
* Close: https://github.com/mapbox/mapbox-navigation-ios/issues/1020

This changes the arrival event to fire whenever a user arrives at a waypoint instead of just the final waypoint.

/cc @mapbox/navigation-ios @ericrwolfe

### CODE REVIEW

File: MapboxCoreNavigation/MMEEventsManager.swift [hunk 1]
```diff
@@ -192,6 +192,8 @@ extension MMEEventsManager {
     func navigationArriveEvent(routeController: RouteController) -> [String: Any] {
         var eventDictionary = self.addDefaultEvents(routeController: routeController)
         eventDictionary["event"] = MMEEventTypeNavigationArrive
+        eventDictionary["totalDestinationWaypoints"] = routeController.routeProgress.route.legs.count
+        eventDictionary["remainingDestinationWaypoints"] = routeController.routeProgress.remainingWaypoints.count
```
Comment (bsudekum) [inline MapboxCoreNavigation/MMEEventsManager.swift]: @ericrwolfe are you down with these key additions and their implications?

Comment (ericrwolfe) [reply MapboxCoreNavigation/MMEEventsManager.swift]: Yeah, makes sense, could you also be sure to update the spec + increment the events version if you add/change keys?

Also, would it make sense to add these to all the events? Or rename to `currentLegIndex`, `currentStepIndex`, `totalLegCount`, `totalStepCount`?

Comment (bsudekum) [reply MapboxCoreNavigation/MMEEventsManager.swift]: Yeah, let's make these changes.

File: MapboxCoreNavigation/MMEEventsManager.swift [hunk 2]
```diff
@@ -111,6 +116,13 @@ struct EventDetails {
             totalTimeInBackground += abs(session.lastTimeInBackground.timeIntervalSinceNow)
         }
         percentTimeInForeground = totalTimeInPortrait + totalTimeInLandscape == 0 ? 100 : Int((totalTimeInPortrait / (totalTimeInPortrait + totalTimeInLandscape) * 100))
+        currentLegIndex = routeController.routeProgress.legIndex
+        totalLegCount = routeController.routeProgress.route.legs.count
+        currentStepIndex = routeController.routeProgress.currentLegProgress.stepIndex
+        totalStepCount = routeController.routeProgress.currentLeg.steps.count
```
Comment (ericrwolfe) [inline MapboxCoreNavigation/MMEEventsManager.swift]: Let's change these keys to:
- `stepIndex` (index of current step on current leg)
- `stepCount` (count of steps on the current leg)
- `legIndex` (index of leg in current route)
- `legCount` (count of legs on the current route)
- `totalStepCount` (sum of all steps in the route)

File: MapboxCoreNavigation/MMEEventsManager.swift [hunk 3]
```diff
@@ -111,6 +116,13 @@ struct EventDetails {
...
+        
+        let maneuver = routeController.routeProgress.currentLegProgress.upComingStep ?? routeController.routeProgress.currentLegProgress.currentStep
+        maneuverHash = "\(maneuver.names?.joined(separator: ",") ?? "")-\(maneuver.destinations?.joined(separator: ",") ?? "")-\(maneuver.codes?.joined(separator: ",") ?? "")-\(maneuver.initialHeading ?? 0)-\(maneuver.finalHeading ?? 0)"
```
Comment (ericrwolfe) [inline MapboxCoreNavigation/MMEEventsManager.swift]: Don't worry about combining this yourself. We can keep these as separate fields and join later during analysis.

We can add these in a follow on PR if you want to keep this scoped to just the index additions.

File: MapboxCoreNavigation/RouteController.swift [hunk 4]
```diff
@@ -527,6 +533,13 @@ extension RouteController: CLLocationManagerDelegate {
         
         let filteredLocations = locations.filter { $0.isQualified }
         
+        // Always update no matter if location is bad
+        if let lastLocation = filteredLocations.last {
```
Comment (ericrwolfe) [inline MapboxCoreNavigation/RouteController.swift]: Is this first if statement necessary if you have the second?

File: MapboxCoreNavigation/RouteController.swift [hunk 5]
```diff
@@ -527,6 +533,13 @@ extension RouteController: CLLocationManagerDelegate {
...
+            sessionState.pastLocations.push(lastLocation)
+        } else if let lastLocation = locations.last {
```
Comment (ericrwolfe) [inline MapboxCoreNavigation/RouteController.swift]: a) Should we exit early if there is no final location?
b) Should we just add all the locations in the array?

File: MapboxCoreNavigation/MMEEventsManager.swift (hunk 2 above)
Comment (bsudekum) [reply MapboxCoreNavigation/MMEEventsManager.swift]: Noting we already have a `stepCount` variable.

File: MapboxCoreNavigation/RouteController.swift (hunk 5 above)
Comment (bsudekum) [reply MapboxCoreNavigation/RouteController.swift]: Yeah let's do b.

**Comment by bsudekum** [comment]
@ericrwolfe ok, this is ready for another round of review.

**Comment by bsudekum** [comment]
@mapbox/navigation-ios this is ready for review.

File: MapboxCoreNavigation/MMEEventsManager.swift [hunk 6]
```diff
@@ -74,7 +80,7 @@ struct EventDetails {
             self.geometry = Polyline(coordinates: geometry)
             distance = round(session.currentRoute.distance)
             estimatedDuration = round(session.currentRoute.expectedTravelTime)
-            stepCount = session.currentRoute.legs.map({$0.steps.count}).reduce(0, +)
+            stepCountRemaining = session.currentRoute.legs.map({$0.steps.count}).reduce(0, +)
```
Comment (frederoni) [inline MapboxCoreNavigation/MMEEventsManager.swift]: nit: () can be omitted on trailing closures

File: MapboxCoreNavigation/MMEEventsManager.swift [hunk 7]
```diff
@@ -171,6 +183,12 @@ struct EventDetails {
         
         modifiedEventDictionary["percentTimeInPortrait"] = percentTimeInPortrait
         modifiedEventDictionary["percentTimeInForeground"] = percentTimeInForeground
+        
+        modifiedEventDictionary["stepIndex"] = stepIndex
+        modifiedEventDictionary["stepCount"] = totalStepCount
```
Comment (ericrwolfe) [inline MapboxCoreNavigation/MMEEventsManager.swift]: `totalStepCount` is shared here between `modifiedEventDictionary["stepCount"]` and `modifiedEventDictionary["totalStepCount"]`

**Pull request merged**